File: platforms/android/src/main/kotlin/com/siansiansu/taigikeyboard/emojis/TaigiEmojiStore.kt
// NOTE: Not shared-core — Android platform loader (uses android Context + org.json).
// The value types in TaigiEmoji.kt are the shared-core-clean half.

package com.siansiansu.taigikeyboard.emojis

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.InputStream

object TaigiEmojiStore {
    private const val ASSET_PATH = "emoji.json"

    /** Load the emoji document bundled in this library's assets. */
    fun load(context: Context): TaigiEmojiDocument =
        context.assets.open(ASSET_PATH).use(::decode)

    fun decode(input: InputStream): TaigiEmojiDocument {
        val root = JSONObject(input.readBytes().decodeToString())
        val metaObject = root.getJSONObject("meta")
        val meta = TaigiEmojiDocument.Meta(
            emojiVersion = metaObject.getString("emojiVersion"),
            cldrVersion = metaObject.getString("cldrVersion"),
            count = metaObject.getInt("count"),
        )
        val categories = root.getJSONArray("categories").objects().map { categoryObject ->
            TaigiEmojiCategory(
                id = categoryObject.getString("id"),
                title = categoryObject.getString("title"),
                order = categoryObject.getInt("order"),
                emoji = categoryObject.getJSONArray("emoji").objects().map(::decodeEmoji),
            )
        }
        return TaigiEmojiDocument(meta, categories)
    }

    /** Emoji whose merged keyword list contains [query] (case-insensitive substring). */
    fun search(query: String, document: TaigiEmojiDocument): List<TaigiEmoji> {
        val needle = query.lowercase()
        if (needle.isEmpty()) return emptyList()
        return document.categories
            .asSequence()
            .flatMap { it.emoji.asSequence() }
            .filter { emoji -> emoji.keywords.any { it.contains(needle) } }
            .toList()
    }

    /**
     * Drop emoji the host OS font cannot render. The caller supplies the check (e.g.
     * `PaintCompat.hasGlyph`). Test the WHOLE [TaigiEmoji.base] string, not per-scalar —
     * an unsupported ZWJ sequence can falsely pass a per-codepoint check.
     */
    fun filteringUnrenderable(
        document: TaigiEmojiDocument,
        isRenderable: (String) -> Boolean,
    ): TaigiEmojiDocument {
        val categories = document.categories
            .map { category -> category.copy(emoji = category.emoji.filter { isRenderable(it.base) }) }
            .filter { it.emoji.isNotEmpty() }
        return document.copy(categories = categories)
    }

    private fun decodeEmoji(json: JSONObject): TaigiEmoji {
        val locale = json.getJSONObject("keywordsByLocale")
        return TaigiEmoji(
            base = json.getString("base"),
            cp = json.getJSONArray("cp").ints(),
            name = json.getString("name"),
            subgroup = json.getString("subgroup"),
            version = json.getString("version"),
            variations = json.getJSONArray("variations").strings(),
            keywords = json.getJSONArray("keywords").strings(),
            keywordsByLocale = TaigiEmoji.KeywordsByLocale(
                en = locale.getJSONArray("en").strings(),
                zhHant = locale.getJSONArray("zh_Hant").strings(),
                taigi = locale.getJSONArray("taigi").strings(),
            ),
        )
    }

    private fun JSONArray.objects(): List<JSONObject> = (0 until length()).map { getJSONObject(it) }

    private fun JSONArray.strings(): List<String> = (0 until length()).map { getString(it) }

    private fun JSONArray.ints(): List<Int> = (0 until length()).map { getInt(it) }
}

File: platforms/apple/Sources/TaigiEmojis/TaigiEmojiStore.swift
// Loads + searches the bundled emoji document. Foundation-only so it runs under
// `swift test` on macOS; glyph rendering (UIKit/CoreText) is injected by the host app.

import Foundation

public enum TaigiEmojiError: Error {
    case resourceMissing
}

public enum TaigiEmojiStore {
    /// Decode the emoji.json bundled with this package.
    public static func load() throws -> TaigiEmojiDocument {
        guard let url = Bundle.module.url(forResource: "emoji", withExtension: "json") else {
            throw TaigiEmojiError.resourceMissing
        }
        return try JSONDecoder().decode(TaigiEmojiDocument.self, from: Data(contentsOf: url))
    }

    /// Decode an emoji document from raw JSON (e.g. a custom bundle or remote payload).
    public static func decode(from data: Data) throws -> TaigiEmojiDocument {
        try JSONDecoder().decode(TaigiEmojiDocument.self, from: data)
    }

    /// Emoji whose merged keyword list contains `query` (case-insensitive substring).
    public static func search(_ query: String, in document: TaigiEmojiDocument) -> [TaigiEmoji] {
        let needle = query.lowercased()
        guard !needle.isEmpty else { return [] }
        return document.categories
            .flatMap(\.emoji)
            .filter { emoji in emoji.keywords.contains { $0.contains(needle) } }
    }

    /// Drop emoji the host OS font cannot render. The caller supplies the check because
    /// CoreText/UIKit live in the app, not this package. Test the WHOLE `base` string,
    /// not per-scalar — an unsupported ZWJ sequence can falsely pass a per-codepoint check.
    public static func filteringUnrenderable(
        _ document: TaigiEmojiDocument,
        isRenderable: (String) -> Bool
    ) -> TaigiEmojiDocument {
        let categories = document.categories
            .map { category in
                TaigiEmojiCategory(
                    id: category.id,
                    title: category.title,
                    order: category.order,
                    emoji: category.emoji.filter { isRenderable($0.base) }
                )
            }
            .filter { !$0.emoji.isEmpty }
        return TaigiEmojiDocument(meta: document.meta, categories: categories)
    }
}
